### PR TITLE
Fix one annoying and one serious concurrency bug

### PR DESCRIPTION
**Pull request by tbg**
Annoying:
We had a performance bug which caused the MaxTimestamp to be set to the
Timestamp instead of OrigTimestamp on txn reads, which always left the
interval (OrigTimestamp, Timestamp) open to read uncertainty errors.

Serious:
On txn reads (which must always be carried out with the original timestamp
of the transaction, OrigTimestamp), an old bit of code left over from a
larger refactor forwarded the timestamp to the transaction's current commit
timestamp. Perplexingly, this was only caught recently in a manual shell
session. The fix was trivial and a test has been added. Fixes #4293.

The two items are in a single commit because the first phenomenon happens
only after you fix the second, and writing the test for the second needs the fix
for the first.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4332)

### CODE REVIEW

File: storage/replica.go [hunk 1]
```diff
@@ -1466,32 +1466,19 @@ func (r *Replica) executeBatch(batch engine.Engine, ms *engine.MVCCStats, ba roa
...
-			if fiddleWithTimestamps && args.Method() != roachpb.PushTxn {
-				ts = ba.Timestamp.Add(0, int32(index))
-			} else if !isTxn {
-				ts = ba.Timestamp
-			} else {
-				// TODO(tschottdorf): it can happen that the request timestamp
-				// is ahead of the Txn timestamp. A few tests do this on pur-
-				// pose; should undo that and see if there's a legitimate
-				// reason left to keep allowing this behavior.
-				ts.Forward(ba.Txn.Timestamp)
```
Comment (tschottdorf) [inline storage/replica.go]: This is the "serious bug", authored by yours truly sometime last year.

**Comment by RaduBerinde** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/testdata/snapshot_certain_read, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAGz8Ab4t5-C2gZfx6D:-KAGz8Ab4t5-C2gZfx6E:-1965546791) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_certain_read#L46)):*
Some statements are `INTO t` some are `INTO t(a)`. I would make them all the same

---

*[sql/testdata/snapshot_certain_read, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAGzOUsELkPaz6AYeGP:-KAGzOUsELkPaz6AYeGQ:-1628550912) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_certain_read#L49)):*
maybe mention the issue number for the bug

---

*[sql/testdata/snapshot_certain_read, line 51 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAGzX8m-HogqqmR1YSf:-KAGzX8m-HogqqmR1YSg:-583459464) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_certain_read#L51)):*
"the read below". also is "reading the value 2" supposed to be part of the sentence?

---

*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAH-LA34mcOVKWomH9v:1985680082) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
What's the difference between these two tests? They seem the same (other than the comments)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/stores.go, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH52UUT_VwPkfdvlZk:-KAH52UUT_VwPkfdvlZl:-984947203) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L164)):*
Do you need to change this?

---

*[storage/stores.go, line 173 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH2465UaeiVZ8KvGZY:-KAH2465UaeiVZ8KvGZZ:1028193018) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L173)):*
s/avoidable/unnecessary/

---

*[storage/stores.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH2sgYlsX_BCm-7XO8:-KAH2sgYlsX_BCm-7XO9:-1950827111) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L177)):*
You are setting the MaxTimestamp to be behind the Timestamp. That seems wrong.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by tschottdorf** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/testdata/snapshot_certain_read, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAGzOUsELkPaz6AYeGP:-KAH1AoTJTOOwjxdGFrR:-1756286184) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_certain_read#L49)):*
No issue, found while fixing the serious bug.

---

*[sql/testdata/snapshot_certain_read, line 51 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAGzX8m-HogqqmR1YSf:-KAH1FMuS8EgnNunqXLg:-1900605100) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_certain_read#L51)):*
No, fixed.

---

*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAH3G5goQEDl-BelLUx:1685751430) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
Yes, they're the same because they both exercise both bugs. That looks weird in review, but I think clarity would suffer if I tried to explain the two otherwise orthogonal failure modes in one. 

---

*[storage/stores.go, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH52UUT_VwPkfdvlZk:-KAH6K1-8-BIY2whIMCT:-569400455) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L164)):*
Yep, done.

---

*[storage/stores.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH2sgYlsX_BCm-7XO8:-KAH74CbYGI3ubtz2Edx:-835024377) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L177)):*
No, that's correct. MaxTimestamp only has an effect when it's ahead of the active timestamp, which is either `Timestamp` or `OrigTimestamp`. Perhaps a better name for `MaxTimestamp` would be `MaxTimestampValuesBeforeWhichAreWithinOurUncertaintyInterval`, but I hope the field definition has a descriptive comment already.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by RaduBerinde** [comment]
*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAHF2iuKyh_RaOXv3MH:962560570) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
Both failure modes are related to the final SELECT *.. You could describe the two bugs there:
```
# We check for two failure modes here:
#
#  - we must not run into clock uncertainty (Issue #xx) Bla bla
#
#  - we must not read transactionB's uncommited data (Issue #yy) Ramble ramble
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-:-KAHI6gPKl_TZSIJaB_D:676834233)*

**Comment by tschottdorf** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAHNulpw_2vMQhIouLc:797918971) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
The failures are quite orthogonal. One test fails because it's encountering a value in the future and restarts when it shouldn't. The other test failure is about reading data it's not even supposed to see. There's no relation except that I happened to find code which tickles both bugs at the same time. I can understand that it tickles your OCD that there would be byte-identical tests (mod comments) but they run fast, and I think the reasoning is going to be hard to follow with two comment threads going on and less descriptive names. If you feel as strongly as I do, let's do a round of online RPS (but the winner has to make the change) :-)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by RaduBerinde** [comment]
LGTM

---

Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAHPIyGRPfC3avrlbvP:1546025502) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
I guess the thing that bugs me is that if either thing breaks, both tests would fail anyway (so at least with respect to the tests, the failures are not so orthogonal). But I don't feel strongly about it, this is ok to merge. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-:-KAHPI2k0XMiHT12fC4u:676834233)*

**Comment by tschottdorf** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/testdata/snapshot_timestamp_drift, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH-LA34mcOVKWomH9u:-KAHSIw3XIFGAnjwMVuy:-1428712332) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/sql/testdata/snapshot_timestamp_drift#L1)):*
That's a valid way of looking at it. Some rejiggling of the tests would change that (could write into the future with a low prio and not commit in the uncertainty test), but I'm going to leave as is (anticipating that at least the drifting snapshot test will never, ever, ever fail again). I already looked up an online RPS side. Too bad, maybe another time.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/stores.go, line 177 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332#-KAH2sgYlsX_BCm-7XO8:-KAI5i9FprJb_H6OGjR9:2092384006) ([raw file](https://github.com/cockroachdb/cockroach/blob/c53092975e2d11d8359299ec2f123c9a4d86d09a/storage/stores.go#L177)):*
Once this is checked in, can you update the design doc. Thanks

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4332)*

**Comment by tschottdorf** [comment]
@vivekmenezes there's nothing to update afaict:

> Additionally, the conflicting node is
marked as “certain”. Then, for future reads to that node within the
transaction, we set `MaxTimestamp = Read Timestamp`, preventing further
uncertainty restarts.
